tests/api2/test_040_ad_user_group_cache.py

```diff
@@ -0,0 +1,322 @@
+    LDAP / AD users.
+    """
+    results = GET('/directoryservices/cache_refresh/')
+    assert results.status_code == 200, results.text
+    if results.status_code == 200:
+        refresh_job = results.json()
+        job_status = wait_on_job(refresh_job, 180)
+        assert job_status['state'] == 'SUCCESS', str(job_status['results'])
+        if job_status['state'] == 'SUCCESS':
+            rebuild_ok = True
+
+    """
+    Verify that the AD user / group cache was rebuilt successfully.
+    """
+    if rebuild_ok:
+        results = GET('/group', payload={
+            'query-filters': [['local', '=', False]],
+            'query-options': {'extra': {"search_dscache": True}},
+        })
+        assert results.status_code == 200, results.text
+        assert len(results.json()) > 0, results.text
+
+        results = GET('/user', payload={
+            'query-filters': [['local', '=', False]],
+            'query-options': {'extra': {"search_dscache": True}},
+        })
+        assert results.status_code == 200, results.text
+        assert len(results.json()) > 0, results.text
+
+
+@pytest.mark.dependency(name="LAZY_INITIALIZATION_BY_NAME")
+def test_10_check_lazy_initialization_of_users_and_groups_by_name(request):
+    """
+    When users explicitly search for a directory service or other user
+    by name or id we should hit pwd and grp modules and synthesize a
+    result if the user / group is not in the cache. This special behavior
+    only occurs when single filter of "name =" or "id =". So after the
+    initial query that should result in insertion, we add a second filter
+    to only hit the cache.
+    """
+    depends(request, ["pool_04", "REBUILD_AD_CACHE"], scope="session")
+    global ad_user_id
+    global ad_domain_users_id
+    domain_prefix = f'{WORKGROUP.upper()}{WINBIND_SEPARATOR}'
+    ad_user = f'{domain_prefix}{ADUSERNAME.lower()}'
+    ad_group = f'{domain_prefix}domain users'
+
+    cmd = 'rm -f /root/tdb/persistent/*'
+    results = SSH_TEST(cmd, user, password, ip)
+    assert results['result'] is True, results['output']
+
+    if not results['result']:
+        return
+
+    results = GET('/user', payload={
+        'query-filters': [['username', '=', ad_user]],
+        'query-options': {'extra': {"search_dscache": True}},
+    })
+    assert results.status_code == 200, results.text
+    assert len(results.json()) > 0, results.text
+    if len(results.json()) == 0:
+        return
+
+    ad_user_id = results.json()[0]['uid']
+    assert results.json()[0]['username'] == ad_user, results.text 
+
+    results = GET('/group', payload={
+        'query-filters': [['name', '=', ad_group]],
+        'query-options': {'extra': {"search_dscache": True}},
+    })
+    assert results.status_code == 200, results.text
+    assert len(results.json()) > 0, results.text
+    if len(results.json()) == 0:
+        return
+
+    ad_domain_users_id = results.json()[0]['gid']
```

Comment:
`ad_domian_users_id` is set, but the `assert()` below is looking at `ad_group` ... same thing ?